Hacky fixes to archive likes

aspensmonster opened the pull request:
This doesn't actually have to be merged. It's more for anyone else who's looking to archive their years worth of likes before Tumblr `rm -rf`s them into oblivion.

I intended to make it cleaner, but given Tumblr's two week deadline before they delete all NSFW content --and I'm sure plenty of other content that's going to get swept up in this accidentally-- I figured someone might find it useful in its present state.

The code is updated to archive all likes, with a ten second pause between API calls to try and avoid hitting API quotas. The previous code would only get the 1000 most recent likes.

Potential remaining problems:

  * I don't think I've gotten incremental backup working with this yet (can't just rely on latest timestamp like you can for a blog's posts, since a user could conceivably run around "liking" older posts; I think that's what stopped me from trying to tackle this before, since I'd have to iterate over the whole collection anyway to make sure nothing was missed).
  * The rendered HTML shows the chronological order of the content's posting, not the time when the user actually "liked" the post. 
  * No way to view likes by tag (though it looks like this is an open feature request)

I was going to get around to those problems "eventually", but my philosophy is that I can also download the json payloads and sort it all out... "eventually".

I've been running this script like so:

```
./tumblr_backup.py --likes --outdir=/home/some/path/tumblr_backup_likes/ --save-video --save-audio -j --exif='' some_blog_name
```

_Edit: Be sure that the outdir is **not** where you have your own blog's posts backed up. You'll overwrite the html code. As far as this script is concerned, a backup of a blog's posts, and a backup of a blog's likes, both render to the same HTML layout. So if you want to keep both, you need to separate the two in different paths._

I've saved 70 GiB worth of historical likes this way. Not everything is saved. Some posts caused youtube-dl to choke out (an annoying problem that never goes away given the frenetic update cadence that most video sites seem to adhere to). Some videos that are downloaded don't seem to want to play. And of course some of the liked content has been deleted over the years. But 70 GiB of mostly-good salvage of 30k+ likes is better than 0.

Obviously the likes have to be public, and if you want to gather everything --including stuff that has been NSFW tagged? Or something else? All I know is my original run didn't grab everything-- then the blog will need to, for the time of running the script at least, be marked as containing sensitive content. Or something. I don't remember all of the specifics. Just that a second pass after doing that yielded more results.

Also, I'd advise setting up your own app and using your own API key. Likes can only be iterated over 20 at a time, and if you have tens of thousands of likes, then you could conceivably go over hourly/daily limits. I don't know what quotas the public API key that the script is using has (maybe it's not rate limited?), but it'd probably be best to be a good neighbor and get your own API key to use with these tweaks.

aspensmonster committed:
Hacky fixes to archive likes

vl09 commented:
You're a legend!!! Just saved a few thousand posts and likes before everything goes down. You only forgot to define MAX_LIKES there but it works perfectly fine otherwise! Thank you!!

Hrxn commented:
Woah, wait a sec..

I've heard about the *issues* Tumblr recently had with its App on Apple's Walled Garden (App Store), but this is new to me:
> [..]  but given Tumblr's two week deadline before they delete all NSFW content  [..]

Is this official?

aspensmonster committed:
Include MAX_LIKES definition

aggroskater commented:
> You only forgot to define MAX_LIKES there but it works perfectly fine otherwise! Thank you!!

Because of course I did. Didn't pay quite enough attention when `git add -p`ing the change (didn't want to spill my app api key accidentally). It's there now.

aggroskater commented:
@Hrxn 

From this link:

https://staff.tumblr.com/post/180758987165/a-better-more-positive-tumblr

> **So what’s next?**
>
>Starting December 17, 2018, we will begin enforcing this new policy. Community members with content that is no longer permitted on Tumblr will get a heads up from us in advance and steps they can take to appeal or preserve their content outside the community if they so choose. All changes won’t happen overnight as something of this complexity takes time.
>
>Another thing, filtering this type of content versus say, a political protest with nudity or the statue of David, is not simple at scale. We’re relying on automated tools to identify adult content and humans to help train and keep our systems in check. We know there will be mistakes, but we’ve done our best to create and enforce a policy that acknowledges the breadth of expression we see in the community.

I'm reading "steps they can take to appeal or preserve their content **outside the community**" to mean that, eventually --perhaps not on the 17th two weeks from now, but eventually-- the content will be deleted. And yes, this kind of filtering "is not simple at scale." I'd argue it's not possible at scale. There's already plenty that's getting mis-flagged.

Hrxn commented:
Damn. Thanks. Agreed, I read it in the same way. A real shame.

Laydmei commented:
A little question from a beginner:
I tried the command _"./tumblr_backup.py --likes --outdir=/home/some/path/tumblr_backup_likes/ --save-video --save-audio -j --exif=' some_blog_name'"_ but it does not work. How can I download all my likes? Because I can not download more than 1000 likes. ?_?

vl09 commented:
> A little question from a beginner:
> I tried the command _"./tumblr_backup.py --likes --outdir=/home/some/path/tumblr_backup_likes/ --save-video --save-audio -j --exif=' some_blog_name'"_ but it does not work. How can I download all my likes? Because I can not download more than 1000 likes. ?_?

Be sure to make your blog explicit first in the settings. Somehow, the code works better that way.

Cebtenzzre commented:
It "works better" because AFAIK a non-explicit blog will not publicly show explicit likes.

Laydmei commented:
I have checked. I tried again. But it still does not download all my Likes. Maybe I have too many Likes? I really do not understand.

Doty1154 commented:
Huh, yeah i still get rate limited after the first couple hundred likes. i got like 40 gig down so far. Cries

Laydmei commented:
An idea of how to solve the problem? Please ?

Hrxn commented:
Did you make the code changes (as seen under *Files changed*)?

Cebtenzzre commented:
@Hrxn Lol, manually patching?  Why not just clone [aggroskater/tumblr-utils](https://github.com/aggroskater/tumblr-utils)?

KlfJoat commented:
There's a way to overcome the 10 per run limit.  But you need to use the before/after params differently.  Like, pull the timestamp of the last post you downloaded, and use that in the after param.

Hrxn commented:
Yes, obviously.
I meant to ask to verify if you are running indeed the changed version.
*Did you verify the code changes..*

KlfJoat commented:
I just added this and it works. Great job!

allefeld commented:
@aggroskater I used your version, and it works much better for likes than the original, thank you!

However, I still only successfully downloaded 5593 of 9626 likes, i.e. I'm missing 4033.

While downloading, the program listed 44 times "HTTP Error 403: Forbidden", most of them with URLs referring to the domain `vtt.tumblr.com`.  
It listed 7 times "HTTP Error 404: Not Found", all of them referring to the domain `66.media.tumblr.com`.  
It listed 6 times "WARNING: Could not send HEAD request" with an URL beginning with `https://www.tumblr.com/privacy/consent?redirect=`. In these cases, messages followed saying that the program is "Falling back on generic information extractor." and "Unable to download video".

I understand that these 44 + 7 + 6 = 57 likes may simply be inaccessible (I tried some of the URLs manually and could verify that), but that accounts only for a tiny fraction of the 4033 likes that were skipped without warning.

Is there any chance you can fix this? If I can do something to help, please tell me.

adamamyl commented:
If your behaviour is like what I've seen elsewhere with the `403` errors on videos, I _think_ it is that the videos no longer exist / are not accessible anymore – i.e. it's not a problem with this script or youtube-dl.

adamamyl commented:
> While downloading, the program listed 44 times "HTTP Error 403: Forbidden", most of them with URLs referring to the domain `vtt.tumblr.com`.

@allefeld If your behaviour is like what I've seen elsewhere with `403` errors on videos, I think it is that the videos no longer exist / are not accessible anymore – i.e. it's not a problem with this script or youtube-dl.

Hrxn commented:
A lot of videos are indeed 'removed', or to be more exact, blocked. That is, `403` is the expected result in those cases.
This was the big Tumblr video purge a while ago..

allefeld commented:
@adamamyl @Hrxn you're probably right with your comments. However, that doesn't explain 4033 - 57 = 3976 likes aren't downloaded and don't have a corresponding error message.

Doty1154 commented:
If only it was possible just the download all of https://www.tumblr.com/likes while logged in via cli,

allefeld commented:
I now think these inconsistencies have nothing to do with @bbolli's or @aggroskater's code, but with the extremely weird and unreliable tumblr API.

I experimented a bit with the API myself. I went through the list first using the query/next value for the next request, and found it skips over likes. I then used the field `liked_timestamp` from the last returned post, that worked a little better. I experimented with the `limit` parameter, and found that for a small value, resulting in a lot of requests, at some point the API simply starts to return 0 posts, even though I know that the requested time point has many likes before it. Mind you, no error message, just an "OK" response containing zero posts.

I used  bbolli's code, aggrokaters's fork, https://github.com/javierarce/tumblr-liked-photos-export, and my own code, and I never arrive at the 9000+ likes, get different numbers of recovered posts each time and on different runs.

I'll be turning my back on tumblr soon, and just wanted to get my stuff out before the impending apocalypse. After the both social and technical blunders they commit, I have to say: Good riddance.

Sorry for venting. Thank you for your work!

aggroskater commented:
I'll take a stab at incorporating @qtwyeuritoiy's work into my fork.

aggroskater commented:
I've got the pieces initially merged. It'll take a few hours to do a full grab and then test after liking some other posts.

Sidenote, it seems that the "mark as sensitive" feature, or whatever, is... no longer available in the desktop website's settings. I can't find it anywhere. That also might be playing havoc with downloading likes for all I know. Might break down and try the oauth approach at some point tonight/tomorrow. But that's its own can of worms that'll entail pulling in some library that can support oauth1.0a's HMAC signing mechanism on the requests.

qtwyeuritoiy commented:
@aggroskater Yes, I never touched the incremental backup part so it makes sense it doesn't operate on liked_timestamp. Great job on making the dates work. Another thing I want to see incorporated here is a original blog name in the header. It's a few simple fixes:

```
class TumblrPost:
    def __init__(self, post):
+       self.creator = post['blog_name']

    def get_post(self):
-       post += u'<header>\n<p><time datetime=%s>%s</time>\n' % (self.isodate, strftime('%x %X', self.tm))
+       post += u'<header>'
+       if options.likes:
+           post += u'<p><a href=\"http://{0}.tumblr.com/\" class=\"tumblr_blog\">{0}</a>:</p>\n'.format(self.creator)
+       post += u'\n<p><time datetime=%s>%s</time>\n' % (self.isodate, strftime('%x %X', self.tm))
```

aggroskater commented:
:+1:  @qtwyeuritoiy  Yep :D I have that bit in there too in my local code changes.

aggroskater commented:
My code for getting the latest timestamp was off --`strftime('%s')` passes through to local implementation which uses local TZ instead of UTC-- but after patching that up, it looks like a full backup and incremental backup are working. This explains my earlier observations of latest likes "taking forever". Shows me for blaming the API instead of myself... As of now, I can typically sync the latest likes with `--incremental` within a couple minutes of liking at most.

I've got some final testing to do, but I'm reasonably confident that things are working. Going to push my latest changes to my fork now. I'll try to rebase on top of upstream's latest changes at some point tomorrow.

aspensmonster committed:
Support incremental backup, render likes in order of like_timestamp, add OP link

allefeld commented:
@aggroskater, I downloaded your version of `tumblr-backup.py` today (2018-12-11) and tried to use it incrementally (first complete download, then add new likes), and got the following error message:

```
$ tumblr_backup.py --dirs --save-video --save-audio --likes --outdir=likes -i xyz
Traceback (most recent call last):
  File "./code/tumblr_backup.py", line 1274, in <module>
    tb.backup(account)
  File "./code/tumblr_backup.py", line 568, in backup
    fh = open(f,'r')
IOError: [Errno 21] Is a directory: u'/home/ca/Store/lab/byetumblr/likes/posts/721441137749'
```
(blog name and post number changed for privacy)

To me it looks like incremental mode isn't working anymore?

aspensmonster committed:
Ensure incremental backup of likes works when `--dirs` option in use

aggroskater commented:
@allefeld  I did a quick and dirty test of the (possible) fix I just pushed to my fork. From what I can see, the incremental backup works when using the `--dirs` option now. Let me know if works for you or not.

allefeld commented:
@aggroskater yes works for me, thanks!

There's a long delay at the beginning, I guess while it checks the posts that are already present. Not really important, but some kind of progress report might be good.

Cebtenzzre commented:
So far this has been working fine for me with MAX_LIKES set to 50.

aggroskater commented:
I'm working on rebasing on upstream, but there are some lingering issues with NPF posts that I'd rather not commingle with this fork yet. In particular issue #167 and #162. I'm not sure if #162 is really fixed yet; I think I've got cases where NPF videos still aren't getting downloaded locally. I might spin off another branch on my fork to try and deal with them. The documentation on NPF is lengthy (https://www.tumblr.com/docs/npf).

As is, the fork is working well enough for backing up likes. I'd also like to figure out an oauth way to do it, but time is running out, honestly.

I have MAX_LIKES at 20 based on the official API documentation. There are varying reports about just how many likes can make it through in a call, but I figure sticking to the official limit should be (somewhat) safer.

aggroskater commented:
I've got an alternate approach for dealing with the NPF stuff built locally. Rather than try to parse the NPF payloads in the `data-npf` attribute --various issues (#167, #172, #179) in this repo indicate that its format is not reliable-- instead I have the code getting the video inline the same way that we get images inline for "text" posts.

Incidentally, in the process of building this, I found that there is an existing bug in the inline image handler. The regex locks onto the `data-orig-src` attribute of an `img` tag instead of the `src` attribute. I've included a small fix for that too (though the real fix is to [never parse html with regex](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454). Doesn't look like it happens super often, but hey, the theory is I shouldn't see *any* image elements with http links.

Assuming I can get a full archive to run to completion, and can get a few other blogs archived that have been mentioned across a couple issues in this repo, I'll push my latest rebased code.

aspensmonster committed:
Get inline videos, fix inline image src attribute bug

aspensmonster committed:
Show message indicating latest like is being calculated

For large archives, this might take a while, since we have to open each
file.

aggroskater commented:
Ok. I've got a full archive to run to completion, no obvious flaws. I'm running once more without all of my messy debug output, but in the meantime, I've pushed my latest changes, rebased on upstream. Hopefully this addresses the various issues mentioned above (videos still not downloading, broken/incomplete tag indices, missing type/subtype in NPF post problems...).

aspensmonster committed:
Merge branch 'master' into master

aggroskater commented:
I've merged current upstream/master into my master (trying to preserve commit references now that they're in upstream commits; otherwise I would have rebased). I'm running another backup, but otherwise I think the code is good enough to merge at this point. At least, my original hang-ups have all been tackled in the past few days:

  * **(fixed by examining post files for latest <time> element, which is time-of-like for `--like` runs)** I don't think I've gotten incremental backup working with this yet (can't just rely on latest timestamp like you can for a blog's posts, since a user could conceivably run around "liking" older posts; I think that's what stopped me from trying to tackle this before, since I'd have to iterate over the whole collection anyway to make sure nothing was missed).
  * **(post date is time-of-like for `--like` runs)** The rendered HTML shows the chronological order of the content's posting, not the time when the user actually "liked" the post.
  * **(implemented upstream)** No way to view likes by tag (though it looks like this is an open feature request)

Besides adding support for archiving likes (both full and incremental, with and without `--dirs` option), it also uses a sha256 hash of the tag to make the folder name for the tag (nothing but letters and numbers, not too long of a filename), which should address issues in #140 and #183. The trade-off is that the folder name isn't "friendly", but I think trying to make friendly representations of some of the crazy tags on tumblr --that also conform to varying filename restrictions put in place by different OS's and filesystems-- is an uphill battle, and if you're only using the rendered HTML tag index, then the sha256 folder name isn't a concern.

@bbolli  Let me know what you think.

aspensmonster committed:
Can't method chain on a string interpolation tuple, it seems

Maybe there's some fancy syntax for it, but I couldn't find it with a
quick google search. As is, I get the following exception:

AttributeError: 'tuple' object has no attribute 'replace'

This commit just makes a temporary variable to avoid that.

aggroskater commented:
Hmm. Might have found a problem, might not. I'll update in a bit.

aspensmonster committed:
Regex is a fickle thing

The presence of the first space in `img [^>]* src` causes not to match
`<img src=` due to the need for two whitespace.

aggroskater commented:
Alright. Two small commits later and it looks like I'm back in business, at least by a cursory glance of the posts/ directory as I'm running another full backup.

aspensmonster committed:
Further regex refinement by @Cebtenzzre

beret commented:
> Be sure to make your blog explicit first in the settings. Somehow, the code works better that way.


> Sidenote, it seems that the "mark as sensitive" feature, or whatever, is... no longer available in the desktop website's settings. I can't find it anywhere. That also might be playing havoc with downloading likes for all I know. Might break down and try the oauth approach at some point tonight/tomorrow. But that's its own can of worms that'll entail pulling in some library that can support oauth1.0a's HMAC signing mechanism on the requests.

I found a way to work around the removal of this from the web UI!

Go to you blog settings eg. <https://www.tumblr.com/settings/blog/YOURUSERNAME>

Just inspect one of the form buttons which is already off (i used 'show only on dashboard') and replace the `for` value with `tumblelog_flagged_as_nsfw` As in picture:
![inspector screenshot 20181215](https://user-images.githubusercontent.com/69595/50043012-202da700-002a-11e9-8474-6a3222aaef46.png)

Then go back to the dash, and toggle the control.
![dashboard corntrols](https://user-images.githubusercontent.com/69595/50043038-7e5a8a00-002a-11e9-9280-8f8b5eeecb2d.png)
It saved (I could see a 200 response from the API) and i was then able to see my full number of posts when querying.

<3

Code-You-Fools commented:
@beret for whatever reason, your hack doesn't seem to be working anymore. Tried it in incognito mode (to unload all extensions, etc) to no avail.

aspensmonster committed:
Merge branch 'master' into master

aggroskater commented:
The toggle option is visible in the android app at least, though it appears non-functional. Toggling "Account > Visibility > $blog_name is explicit" to on doesn't cause the blog to become masked with the explicit screen to non-logged-in users. Exiting the visibility page of the android app and returning to it shows that the setting is back off.

¯\_(ツ)_/¯

Code-You-Fools commented:
Yeah, I think they may have disabled it server-side, meaning nothing we do can change it. For some reason though using the base tumblr_backup gets me far more likes than using this version. Not sure if it uses a different way of accessing the API that lets it download more likes or what, but honestly I think the best bet is to swap to OAuth and use that.

aggroskater commented:
>For some reason though using the base tumblr_backup gets me far more likes than using this version

By "base tumblr_backup", do you mean the tumblr_backup.py upstream? Or my fork? Because my fork was originally inspired by the fact that, as is, the upstream tumblr_backup.py stops at ~1000 likes (due to a tumblr API limitation when using offset parameters on the /likes endpoint).

Code-You-Fools commented:
The upstream one downloads 1471 of my likes, but yours only gets 366 for some reason. The total is 1484 according to both versions and Tumblr's website.

aggroskater commented:
Strange. On my own blog with 36k+ likes I only get around a 1000 with upstream, but with my fork I get 28k. I figure the difference comes in from deleted posts and explicit ones. I can look into it, but I'd need the full command that was run along with the blog name.

aggroskater commented:
Alright. I've run with current upstream master (well, commit 85b4ce3). The output says that it backed up 1500 or so likes, but what actually gets archived is less:

```
$ python2 tumblr_backup.py --save-audio --save-video -l -S -O ~test just-bunbun-things
just-bunbun-things: 1471 posts backed up 
$ ls ~test/posts/ | wc -l
290
```

Code-You-Fools commented:
Huh. Well, guess it's just the script being dumb and claiming it processed more than it did. Still can't mark the blog as explicit sadly, so no way to get the rest of the likes unless there's an OAuth implementation of this. I could probably do it, but I don't have access to an API key that isn't rate-limited, and I doubt that Tumblr will unlimit one of mine by the start of the apocalypse on Monday.

Code-You-Fools commented:
@beret Just checked (changed things inside of the `binary_switch`, which is what controls what is sent to the servers) and it looks like Tumblr disabled flagging your own blog as explicit on their side.

Soundsgoood commented:
Hey guys, trying this out, and it works great so far, surpassing the ~950 files I get from bbolli's version.
I want to try it with --save-audio and --save-video this time, so I've got a question for you all: How do I install youtube-dl for Windows? Putting the .exe from the youtube-dl page in the same folder as the tumblr_backup.py file doesn't work, (as in tumblr_backup.py throws me an error saying youtube_dl is not installed,) even though they're both in a PATH directory.

nightpool commented:
you need to install the youtube_dl python module

On Sun, Dec 16, 2018, 5:07 PM Orion G. <notifications@github.com> wrote:

> Hey guys, trying this out, and it works great so far, surpassing the ~950
> files I get from bbolli's version.
> I want to try it with --save-audio and --save-video this time, so I've got
> a question for you all: How do I install youtube-dl for Windows? Putting
> the .exe from the youtube-dl page in the same folder as the
> tumblr_backup.py file doesn't work, (as in tumblr_backup.py throws me an
> error saying youtube_dl is not installed,) even though they're both in a
> PATH directory.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bbolli/tumblr-utils/pull/114#issuecomment-447680474>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAORVyzKnICJetRuOXlUdWlcVdTnqKJQks5u5sQjgaJpZM4Y_tiQ>
> .
>

aggroskater commented:
According to Tumblr staff's latest post, things aren't getting deleted quite yet, but hidden from view. Perhaps historical explicit likes are still accessible to the user if logged in? I'll try to implement an oauth approach in the coming days to see if that's the case.

Hrxn commented:
@aggroskater Yeah, can confirm. *Sensitive* content is hidden by Tumblr's web front-end (and I assume it is the same for the mobile apps), but the API still returns the same results as before, including URLs to the image files or clips not visible in the browser anymore.
I've been using gallery-dl, which supports Tumblr with OAuth authentication. Without any external dependencies, if I may add.

allefeld commented:
@aggroskater I made another attempt at saving a few more likes today, but now I'm getting an error message:

```
$ ./tumblr_backup.py --dirs --save-video --save-audio --likes --outdir=likes-new blogname
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
HTTP Error 403: Forbidden getting https://api.tumblr.com/v2/blog/blogname.tumblr.com/likes?reblog_info=true&api_key=8YUsKJvcJxo2MDwmWMDiXZGuMuIbeCwuQGP5ZHSEA4jBJPMnJT&limit=1
```
(blog name changed for privacy)

Did they block your API key by any chance?

allefeld commented:
Apparently not... I tried it with my own key and the same thing happens.

allefeld commented:
@Hrxn, I tried gallery-dl, which downloaded the photos from my posts just fine, but I can't figure out how to download likes. Can you give me a hint?

Hrxn commented:
@allefeld Well, gallery-dl determines the selection of the appropriate extractor by matching it against specific URL patterns, this includes support for different variants. I think this should be enough here:
`gallery-dl http(s)://<blog-name>.tumblr.com/likes`

allefeld commented:
> @allefeld Well, gallery-dl determines the selection of the appropriate extractor by matching it against specific URL patterns, this includes support for different variants. I think this should be enough here:
> `gallery-dl http(s)://<blog-name>.tumblr.com/likes`

That's what I thought, but I'm getting

    [tumblr][error] You do not have permission to access the resource at 'https://<blog-name>.tumblr.com/likes'

though I used `-u` and `-p` options. Adding `-v` reveals that gallery-dl does not use OAuth (but simply API key) for likes.

Pity, but thanks!

ytguwn commented:
Unfortunately, it is not possible to fetch all likes even with OAuth. 

I modified gallery-dl to use OAuth secured endpoint (`/user/likes` instead of `/blog/<name>/likes`) and it hasn't brought me anything new.

The same likes are also missing from web interface (`tumblr.com/likes`).

Note: I have ~500 likes on tumblr and currently I am able to grab only half of them (and this number diminishes over time).

allefeld commented:
Well, I guess it's time to say goodbye to tumblr for good...

ytguwn commented:
FWIW there exists an unofficial Twitter SVC API (see #161).

I don't know if it's useful for grabbing likes.

Cebtenzzre commented:
@ytguwn Assuming you mean Tumblr, it probably isn't.  Private likes can be scraped via OAuth (see #200 for something similar), and the likes that aren't found with this probably aren't available via any API (possibly deleted posts).

sldx12 commented:
I get the following error when the script is almost finishing:
```

Traceback (most recent call last):                                         
  File "tumblr_backup.py", line 1335, in <module>
    tb.backup(account)
  File "tumblr_backup.py", line 692, in backup
    before_timestamp = soup['response']['_links']['next']['query_params']['before']
KeyError: 'next'
```

Cebtenzzre commented:
@sldx12 I was able to reproduce this error. The API must've changed because I know this fork worked before. Basically, the API no longer gives the script a pointer to the next (empty) batch of liked posts, but this fork still expects it to be there (and fails when it isn't).
It can be fixed by removing the text ` and not soup['response']['liked_posts']`, moving the try/except block to before the sleep and dedenting it once, and removing the now-empty else clause. I'll update my fork to match.

aggroskater commented:
I'll take a look. I haven't edited my script locally much in the past two years, but it has been working as recently as yesterday.

aggroskater commented:
I just re-ran my script locally. No issues. Judging by the error message, I wonder if the `soup['meta']['status']` bit wasn't 200 at the time of running:

```
                        try:
                            before_timestamp = soup['response']['_links']['next']['query_params']['before']
                        except KeyError:
                            if soup['meta']['status'] == 200 and not soup['response']['liked_posts']:
                                finished_with_likes = True
                                continue
                            else:
                                raise
```

The `except KeyError` bit is meant to handle the lack of `next` element, and the only way I think you'd see a raised `KeyError: 'next'` is if either the meta status wasn't 200 or `soup['response']['liked_posts']` actually *was* present, causing the `raise` in the `else` bit to get hit. I honestly don't recall what the `soup['response']['liked_posts']` bit was about though.

Cebtenzzre commented:
@aggroskater `soup['meta']['status']` is guaranteed to be 200 here, so the check is redundant anyway. If status is non-200, `soup` would be `None` and the `# try the next batch` branch would exit the loop early.

`not soup['response']['liked_posts']` evaluates to true only if the batch is empty, so the code only allows a clean exit if the final empty batch (which would usually hit `if not posts` -> `break`) is the first one to lack a `next` link. In my testing, it seems that (for some of us, at least) the API no longer returns the `next` link in the last non-empty batch. I still get all of my liked posts according to the `liked_count`, so terminating after backing up the first batch that lacks a `next` link doesn't miss anything.

I made a test blog that can reproduce this issue - try `python2 tumblr_backup.py --likes erablenous`, and you should see the exception, assuming your API key works just like sldx12's and mine.

sldx12 commented:
@Cebtenzzre when trying to run your most recent commit I stumble upon the following (I have tried to import it couldn't do it):

```
Traceback (most recent call last):
  File "tumblr_backup.py", line 28, in <module>
    from util import ConnectionFile, HAVE_SSL_CTX, HTTP_TIMEOUT, LockedQueue, PY3, nullcontext, to_bytes, to_unicode
ImportError: No module named util
```

I also tried your fix in 
> It can be fixed by removing the text ` and not soup['response']['liked_posts']`, moving the try/except block to before the sleep and dedenting it once, and removing the now-empty else clause. I'll update my fork to match.
but it gave the following error: 
```
Traceback (most recent call last):                                
  File "tumblr_backup.py", line 1255, in <module>
    tb.backup(account)
  File "tumblr_backup.py", line 645, in backup
    j = min(i + MAX_LIKES, last_post)
NameError: global name 'MAX_LIKES' is not defined
```

aggroskater commented:
@Cebtenzzre I was able to replicate the issue on the test blog you gave with my version of the script. I confirmed that my local script is basically the same as the one at my fork's [master](https://github.com/aggroskater/tumblr-utils/blob/master/tumblr_backup.py), just with a different API key and an added `time.sleep(10)` when backing up posts (not likes; guess I was backing up some big blogs a while back and didn't want to blow through my API key's quota). Merely removing the `and not soup['response']['liked_posts']` permitted the script to complete for me. Now I'm curious as to why I was able to back up my own blog's likes last night without issue. Presumably, the last page of results would be non-empty too (and missing the `next` piece) and I should have hit the same exception. Unless of course the API is behaving differently depending on the blog that's queried :shrug: 

What you're describing is jogging my memory a bit  though. I'm assuming when I first wrote this that the API would always have a `next` link, even on the "last" page of results. So the loop would run once more, the API would return an empty page, no `next` element, and my logic would say "oh, it's a 200 response AND there's no next link AND the list is empty, so really that just means we're done."

I'm going to try again on my own blog after adding only one like. In theory the API should return one page with a single like, no `next` element in the `_links` piece, and fail. If not, well, then I'll at least have the two different API responses from my blog and erablemous to compare and scratch my head with.

aggroskater commented:
Hmmm. I'm guessing that the reason it looks like it's "working" for me is because I was doing an incremental backup of likes, and not a full run. And the like backup is starting from the latest likes and working backward in time. And since I'm doing an incremental backup, I just stop after reaching `ident_max`, well before I get to the first ever like. And in that scenario the `next` element (which is actually providing a `before` timestamp to keep going *backwards* (seriously, what was I thinking when I wrote this damn thing?)) obviously still exists. But for a full run, you get to the earliest set of likes and then... no more `next` element when the last, most early page of results is returned.

I'm guessing at some point maybe the API would have given a "yeah, here's the link to 'before the earliest like ever', it's an empty page, have fun" response. And now it doesn't. Regardless, taking off the `and not soup['response']['liked_posts']` bit on [Line 694](https://github.com/aggroskater/tumblr-utils/blob/master/tumblr_backup.py#L694) still leaves my incremental backups working, and it also backs up the erablenous blog too. So that would be the "fix" if you're rocking this two-year-old script.

But it definitely looks like you've done a significant amount of work on the project, @Cebtenzzre , so I'll probably look into your fork and/or the original @bbolli repo and see if there's a newer, better way for me to keep clinging to my tumblr likes :P

Cebtenzzre commented:
@sldx12 You'll have to download my fork as a zip or use `git clone`, because there are new dependencies like util.py, wget.py, and note_scraper.py. And are you mixing and matching scripts? MAX_LIKES is defined at the top of [aggroskater's version](https://raw.githubusercontent.com/aggroskater/tumblr-utils/master/tumblr_backup.py) (this is the PR you're commenting on), which is the one I meant for you to modify.
@aggroskater Yeah, incremental backups stop early, so they would never see this issue (which happens at the _end_ of the backup). TBH your fork still works better for likes than upstream - the offset parameter that upstream still uses gets stuck in an endless loop (with an unmerged workaround). But please check out my fork, it's full of neat things I've added in the two years I've been using this script, and recently got support for 1000+ likes that works similarly to this PR.

sldx12 commented:
@Cebtenzzre Yeah my bad, I've tried so many stuff that I get confused. I tried your version and it didn't download your likes so I opened a new issue ([here](https://github.com/Cebtenzzre/tumblr-utils/issues/2#issue-712209438)).
As to the MAX_LIKES error, isn't that what you told me to do? I modified the @aggroskater version, not yours.

Cebtenzzre commented:
@sldx12 Basically, there is no way to get that NameError unless you either deleted the line `MAX_LIKES = 20`, or you started from the wrong script. Either way, the line numbers in that traceback don't match the script from this PR at all. They actually match bbolli's version most closely, which would make sense if you got this PR confused with another one, or if you got the script from this repo's homepage instead of the branch this PR is based on.

sldx12 commented:
@Cebtenzzre Ok nevermind, I tried again. I got the same error as with your PR, only with different numbers. So, if you want to reply here or in the issue I created in your PR it's the same for me, I just want it to work.
Error (I have around 1410 liked posts):

`blogname: 1078 posts backed up                                 `

aspensmonster committed:
Archival message
